### PR TITLE
Introduce per object placeholders

### PR DESCRIPTION
**Pull request by tulimaki**
* [x] Add option to do per object placeholders.
* [x] Add tests to prove this works
* [x] Add documenation to all new things
* [x] Add changelog
* [x] Document the new provide
* [x] Add option to pretend anonymous, person or company contact as a shop staff member while editing with xtheme.
* [x] Add browser test for the xtheme editing
* [x] Fix issue with deleting last cell
* [x] Add tests around admin tools
* [ ] Move admin tools styles outside the front styles so updating those would require re-releasing all existing themes.
* [ ] Add some help texts around the visibility limits.

Here we decided to add the visibility limitation to placeholders since that should make it easier for merchant to understand and preview. Also the logic behind "what should be shown for one placeholder" is easier.

Basically now the placeholder can have multiple layouts inside and these different layouts can be provided from apps. Then based on current context the right layouts is shown for each placeholder.

Editing the placeholders is pretty much untouched.

Refs #1220

### CODE REVIEW

File: shuup/xtheme/layout/_base.py [hunk 1]
```diff
@@ -215,6 +217,12 @@ def __init__(self, theme, placeholder_name, rows=None):
         if rows:
             self.rows.extend(rows)
 
+    def is_valid_context(self, context):
+        return True
+
+    def get_layout_data_suffix(self, context):
```
Comment (chessbr) [inline shuup/xtheme/layout/_base.py]: document here for what this is needed

File: shuup/xtheme/layout/_contact_group.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+from shuup.xtheme.layout.utils import get_customer_from_context
+
+
+class AnonymousContactLayout(Layout):
+    identifier = "xtheme-anonymous-contact-layout"
+    help_text = _("Content in this placeholder is shown for guests.")
+
+    def is_valid_context(self, context):
+        customer = get_customer_from_context(context)
+        return (customer.is_anonymous if customer else False)
```
Comment (chessbr) [inline shuup/xtheme/layout/_contact_group.py]: is this test correct? Testing an `AnonymousContact` will always return `False`:
```py
>>> bool(AnonymousContact())
False
```
This way, the context should be valid if customer is `None` of `False` IMO

File: shuup/xtheme/layout/_contact_group.py [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+class PersonContactLayout(Layout):
+    identifier = "xtheme-person-contact-layout"
+    help_text = _("Content in this placeholder is shown for person contacts.")
+
+    def is_valid_context(self, context):
+        customer = get_customer_from_context(context)
+        return (isinstance(customer, PersonContact) if customer else False)
+
+    def get_layout_data_suffix(self, context):
+        return self.identifier
```
Comment (chessbr) [inline shuup/xtheme/layout/_contact_group.py]: maybe a `ContactLayout` should also exist, or even `AuthenticatedUserLayout` that we can show something when there is actually a customer, doesn't matter the type.

File: shuup/xtheme/layout/_category.py [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+# This source code is licensed under the OSL-3.0 license found in the
+# LICENSE file in the root directory of this source tree.
+from django.utils.translation import ugettext_lazy as _
+
+from shuup.xtheme.layout import Layout
+
+
+class CategoryLayout(Layout):
+    identifier = "xtheme-category-layout"
+    help_text = _("Content in this placeholder is shown for this category only.")
```
Comment (chessbr) [inline shuup/xtheme/layout/_category.py]: maybe show the category name here, it could be confusing for user to have 100% what category we are talking about as products can have n categories

**Comment by tulimaki** [comment]
The last 2 steps in todo I am not sure anymore whether those should be done. Even if I move the resources for example the definite theme still have own `base.jinja` and that would still require a release. Also those are just few colors and font size changes.

What comes to the next steps. I think in the future we have "easy" option to group the layouts inside placeholder at front to minimize that the merchant gets confused from so many different layouts. Also the layout could have ordering added while selecting the layout for edit.

File: shuup/front/views/misc.py [hunk 5]
```diff
@@ -7,15 +7,53 @@
...
         return HttpResponseRedirect(return_url)
-    all_seeing_key = "is_all_seeing:%d" % request.user.pk
-    is_all_seeing = not configuration.get(None, all_seeing_key, False)
-    configuration.set(None, all_seeing_key, is_all_seeing)
+
+    toggle_all_seeing_for_user(request.user)
+    return HttpResponseRedirect(return_url)
+
+
+def force_anonymous_contact(request):
```
Comment (chessbr) [inline shuup/front/views/misc.py]: I guess you are repeating yourself with this logic. I guess you could create a decorator for these `is_admin_user` checks

File: shuup_tests/xtheme/test_resources.py [hunk 6]
```diff
@@ -38,7 +38,10 @@ def render(self, context):
         return self.message
 
 
+@pytest.mark.skip()
 def test_resources():
+    # TODO: Give some reason why this test should pass. Why we should
```
Comment (chessbr) [inline shuup_tests/xtheme/test_resources.py]: what's this?

Comment (tulimaki) [reply shuup_tests/xtheme/test_resources.py]: Test is broken since we don't render placeholders if not editable.

Comment (tulimaki) [reply shuup_tests/xtheme/test_resources.py]: I guess point mainly is that should we just remove the test or change it somehow or what?

Comment (chessbr) [reply shuup_tests/xtheme/test_resources.py]: got it. if the user cases doesn't exist, there is no reason for this test then.

File: shuup/front/views/misc.py (hunk 5 above)
Comment (tulimaki) [reply shuup/front/views/misc.py]: Yeah likely, but is it worth. Likely not. And I guess it is not guaranteed that we want same logic to all of these.

File: shuup/core/models/_contacts.py [hunk 7]
```diff
@@ -454,12 +459,31 @@ def get_company_contact(user):
...
+        return None
+
     contact = get_person_contact(user)
     if not contact:
         return None
     return contact.company_memberships.filter(is_active=True).first()
 
 
+def get_company_contact_for_shop(shop):
+    return CompanyContact.objects.get_or_create(
```
[the remark this answers is not in the document]
Comment (tulimaki) [reply shuup/core/models/_contacts.py]: Yeah likely useful to add the shop. But I guess it wouldn't hurt if this would have been more behind the scenes user that is not listed. For now I think I add this.
